Update ray tracing sample to the final ray tracing extensions

SaschaWillems opened the pull request:
## Description

This PR updates the ray tracing sample to use the final cross-vendor ray tracing extensions. It also slightly reworks the code to make it easier to read, adds additional comments to make it easier to follow. 

**Note:** This PR also updates the Vulkan headers and Volk sub modules to Vulkan 1.2.162, which is required for the new extensions.

Due to the changes and split into multiple extensions, the updated sample requires the following extensions:
- VK_KHR_ray_tracing_pipeline
- VK_KHR_acceleration_structure
- VK_KHR_buffer_device_address
- VK_KHR_deferred_host_operations
- VK_EXT_descriptor_indexing
- VK_KHR_spirv_1_4
- VK_KHR_shader_float_controls

It also requires Vulkan 1.1. due to VK_KHR_spirv_1_4.

To make sure it doesn't throw any validation errors or warnings, I have tested this with a recent build of the Validation layers from the source, as the SDK is not yet available. 

**Note**: Trying to run this sample with old validation layers not supporting the new extensions will probably cause errors, so if you want to test with validation, you need to build the layers from source.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

SaschaWillems committed:
Updated ray tracing example to final KHR ray tracing extensions

SaschaWillems committed:
Update Vulkan headers to 1.2.162

SaschaWillems committed:
Update Volk to 1.2.162

SaschaWillems committed:
Merge branch 'master' into raytracing_final

SaschaWillems committed:
Remove no longer required beta compile time flag

SaschaWillems committed:
Set API level to 1.1

This is a requirement for using VK_KHR_spirv_1_4

SaschaWillems committed:
Minor comment wording changes

SaschaWillems committed:
Changed wording in samples readme for the ray tracing sample

JoseEmilio-ARM reviewed (dismissed):
Tested on RTX 2070

gary-sweet commented:
I can confirm that this doesn't break the build or other samples for Broadcom.
I cannot test the new functionality.

marty-johnson59 commented:
Hate to keep leaning on my favorite DT vendors, but as this is DT only and we need quick turn to release the samples next week (to align with the SDK launch), can we get reviews from @mattruschnv and @aurolou?  Thanks!

AaronHaganAMD commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -238,266 +136,377 @@ uint64_t RaytracingBasic::get_buffer_device_address(VkBuffer buffer)
+	const uint32_t           handle_size             = ray_tracing_pipeline_properties.shaderGroupHandleSize;
+	const uint32_t           handle_alignment        = ray_tracing_pipeline_properties.shaderGroupHandleAlignment;
+	const uint32_t           group_count             = static_cast<uint32_t>(shader_groups.size());
+	const uint32_t           sbt_size                = handle_alignment * group_count;
```
I think this should be handle_size * group_count;

SaschaWillems replied:
I'm not sure on this. Wouldn't this result in the buffer being too small on implementations where `shaderGroupHandleSize` is smaller than `shaderGroupHandleAlignment`? I guess we need a function that properly aligns the size instead.

AaronHaganAMD replied:
Thank you for the update, this is passing for us now.

AaronHaganAMD commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -238,266 +136,377 @@ uint64_t RaytracingBasic::get_buffer_device_address(VkBuffer buffer)
+	uint8_t *data = static_cast<uint8_t *>(raygen_shader_binding_table->map());
+	memcpy(data, shader_handle_storage.data(), handle_size);
+	data = static_cast<uint8_t *>(miss_shader_binding_table->map());
+	memcpy(data, shader_handle_storage.data() + handle_alignment, handle_size);
```
I believe this should be handle_size instead of handle_alignment..

SaschaWillems committed:
Align handle sizes where required

SaschaWillems commented:
@AaronHaganAMD: Thanks for reviewing my code. I have pushed a small update that aligns the shader group sizes where required. I can't really test this, as on my implementation the group size is the same as the group size alignment, but I think it should be fine now. It would be great of you could take a look at the changes.

SaschaWillems committed:
Clang format

dgkoch commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -238,266 +136,383 @@ uint64_t RaytracingBasic::get_buffer_device_address(VkBuffer buffer)
+	// Accelerations structures can be build either on the host or the device, with the former only possible if the implementation supports it
+	if (acceleration_structure_features.accelerationStructureHostCommands)
+	{
+		// Implementation supports building acceleration structure building on host
```
Is there any good reason to prefer building on the host (which is optional), instead of just using device build all the time?

SaschaWillems replied:
Not really. Having that path in there may be misleading. Will rework that.

huawei-ahcox replied:
Might be nice to have it controlled from the outside (e.g. commandline) so people can compare build time and quality.

dgkoch replied:
I don't think there is a point for such a basic sample. I'd just keep it simple and only use the guaranteed device path in this sample.

dgkoch commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -238,266 +136,383 @@ uint64_t RaytracingBasic::get_buffer_device_address(VkBuffer buffer)
+	std::vector<VkAccelerationStructureBuildRangeInfoKHR *> acceleration_build_structure_range_infos = {&acceleration_structure_build_range_info};
+
+	// Accelerations structures can be build either on the host or the device, with the former only possible if the implementation supports it
+	if (acceleration_structure_features.accelerationStructureHostCommands)
```
same comment about preferring host builds

SaschaWillems committed:
Always build acceleration structures on the device

Host builds are only mentioned in the comments

SaschaWillems commented:
@dgkoch: I have removed the if-statement and AS are now alway built on the device. I added a comment on host builds, so people know that this is optionally available.

SaschaWillems committed:
Clang format

dgkoch reviewed (approved):
LGTM
